File: sdk/kotlin/src/main/kotlin/app/shadowboy/netplay/sdk/state/CloseReason.kt
package app.shadowboy.netplay.sdk.state

public sealed class NetplayCloseReason {
    public data object Normal : NetplayCloseReason()
    public data object RoomClosed : NetplayCloseReason()
    public data object ReconnectExpired : NetplayCloseReason()
    public data object ProtocolMismatch : NetplayCloseReason()
    public data class RelayError(public val code: String, public val message: String) :
        NetplayCloseReason()
    public data class TransportClosed(public val code: Int? = null, public val reason: String? = null) :
        NetplayCloseReason()
}

File: sdk/typescript/src/state/closeReason.ts
export type NetplayCloseReason =
  | { readonly kind: "normal" }
  | { readonly kind: "roomClosed" }
  | { readonly kind: "reconnectExpired" }
  | { readonly kind: "protocolMismatch" }
  | { readonly kind: "relayError"; readonly code: string; readonly message: string }
  | { readonly kind: "transportClosed"; readonly code?: number; readonly reason?: string };
